### ISSUE TITLE
CPU performance issue: same application has different CPU utilization on XEON, i7 and i3

### ISSUE THREAD

**Issue by divdaisymuffin** [openvinotoolkit/openvino#15994]
hi,

I am trying to otimize CPU on my yolov6 nano model on openvino, I have found out couple of interesting things

When I am using i3 and i7, I am able to run 5 instances of same application on i3, 8 on i7, with 95% CPU usage, while on Xeon silver (6 instances) and Xeon gold only 4, and CPu goes upto 50% max on both Xeon, which I think they have more capacity but somehow they are not taking CPU.

Can you help me with this, because as per Openvino benchmarking Xeon should work better, but in my observation its opposite,

Thanks

**Comment by wangleis** [comment]
@divdaisymuffin By default, OpenVINO will use hyper threading on i3 and i7, but not on Xeon. In your case, hyper threading is enabled on Xeon but not be used.

If you also want to hyper threading on Xeon, please additionally set `ov::num_streams` to the total number of processors on Xeon during `compile_model()`. 

For example `exeNet = core.compile_model(model, "CPU", {ov::num_streams(48)});`

**Comment by dhaval-zala-aivid** [comment]
@wangleis 

For yolov6 we are not using `core.compile_model`

we are doing inference using the below code:
```

self.OVNet = IECore().read_network(
            model = model,
            weights = os.path.splitext(model)[0] + ".bin"
        )
self.OVExec = IECore().load_network(self.OVNet, "CPU")
results = self.OVExec.infer(inputs={self.inputLayer: model_input})
```
let us know how we can set total number of processors here

**Comment by wangleis** [comment]
@dhaval-zala-aivid Please try `self.OVExec = IECore().load_network(self.OVNet, "CPU", {{"CPU_THROUGHPUT_STREAMS", total_number}})`

We recommend developer to use OpenVINO™ API 2.0. Please refer https://docs.openvino.ai/latest/openvino_2_0_transition_guide.html for detail.

**Comment by dhaval-zala-aivid** [comment]
@wangleis 
running yolov8 neno on Xeon Gold using [this script](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/230-yolov8-optimization/230-yolov8-optimization.ipynb) in OpenVINO™ API 2.0 with below code

```
self.compiled_model = core.compile_model(ov_model, device)
core.set_property(device_name="CPU", properties={"NUM_STREAMS":"16"})
```
Here are some strange results:
If I run single instance of yolov8 neno (IR INT8) on video, I'm getting 25 around FPS and it consumes 43% Cpu as we run more instances of yolov8 the FPS get decreased of all the instance, few of them have 2-3 FPS only but CPU usage stuck at  50% only. Why its not using the more cpu to provide the more FPS even after specifying `NUM_STREAMS` ? 

you can see the cpu usage when I'm running 8 instances of yolov8 neno
![image](https://user-images.githubusercontent.com/91198866/222132780-cf931b2a-b0f6-4c0e-910e-ec41fc8c3811.png)

also lscpu: 

![image (1)](https://user-images.githubusercontent.com/91198866/222132858-f80c098e-66a6-44d8-9db5-6bca028317eb.png)

**Comment by wangleis** [comment]
@dhaval-zala-aivid Could you please share the log of `lscpu -e` too? And I updated previous comments, could you please try again?

In your case, `run more instances of yolov8` will load same model multiple times and create multiple model instance for multiple inputs. This is not the correct way to infer multiple inputs with the same network model.

OpenVINO recommend to load network model once and infer multiple inputs with optimal number of infer requests. 

You can use throughput hint to do parallel inference as:
```
auto compiled_model = core.compile_model(model, "CPU", 
    ov::hint::performance_mode(ov::hint::PerformanceMode::THROUGHPUT));
auto num_requests = compiled_model.get_property(ov::optimal_number_of_infer_requests);
```
You also can use all CPU processors for parallel inference as:
```
auto compiled_model = core.compile_model(model, "CPU",  
    ov::hint::performance_mode(ov::hint::PerformanceMode::THROUGHPUT), 
    ov::inference_num_threads(16));
auto num_requests = compiled_model.get_property(ov::optimal_number_of_infer_requests);
```

**Comment by dhaval-zala-aivid** [comment]
Thank you @wangleis for you quick reply 

Eventually I want to run multiple different yolov8n models with different inputs and outputs. But for the testing purpose I'm running same yolov8n model with different inputs and outputs by running the same docker for multiple times.

I have tried to run yolov8n in Asynchronous Mode in python by given code:

```
      ov_model = core.read_model(model) 
      ov_model.reshape([1,3, 416, 416])
      self.input_layer_ir = ov_model.input(0)
      device = "CPU"  # GPU
      if device != "CPU":
          ov_model.reshape({0: [1, 3, 640, 640]})
      config = {"PERFORMANCE_HINT": "THROUGHPUT","INFERENCE_NUM_THREADS": "16"}
      self.compiled_model = core.compile_model(ov_model,device,config)
      self.infer_request = self.compiled_model.create_infer_request()
      .....
      .....
      .....
        input_tensor = image_to_tensor(preprocessed_image)
        self.infer_request.set_tensor(self.input_layer_ir, ov.Tensor(input_tensor))
        self.infer_request.start_async()
        self.infer_request.wait()

        result = self.infer_request.get_output_tensor(0).data
```

Still I didn't notice any improvements in FPS and CPU usage

The inference script is [here](https://drive.google.com/drive/folders/1ta32_564Buz9sKnRv59QJblp6Ivaf7Bb?usp=sharing) with the model as well for your reference

`lscpu -e`

![image](https://user-images.githubusercontent.com/91198866/222357029-e4352239-6377-4447-aa96-95c129181fda.png)

We can also arrange a call as per you availability.

**Comment by wangleis** [comment]
@dhaval-zala-aivid Could you try benchmark_app with throughput hint as below and check CPU behavior? And share the full benchmark app log to me.

- `benchmark_app -m $int8_model_seg_path -d CPU -api async -shape "[1,3,640,640]" -t 15 -hint throughput`

**Comment by dhaval-zala-aivid** [comment]
@wangleis 

with the benchmark cpu has increased to 90%
Logs:

```
[Step 1/11] Parsing and validating input arguments
[ INFO ] Parsing input parameters
[Step 2/11] Loading OpenVINO Runtime
[ INFO ] OpenVINO:
[ INFO ] Build ................................. 2022.3.0-9052-9752fafe8eb-releases/2022/3
[ INFO ]
[ INFO ] Device info:
[ INFO ] CPU
[ INFO ] Build ................................. 2022.3.0-9052-9752fafe8eb-releases/2022/3
[ INFO ]
[ INFO ]
[Step 3/11] Setting device configuration
[Step 4/11] Reading model files
[ INFO ] Loading model files
[ INFO ] Read model took 53.94 ms
[ INFO ] Original model I/O parameters:
[ INFO ] Model inputs:
[ INFO ]     images (node: images) : f32 / [...] / [1,3,?,?]
[ INFO ] Model outputs:
[ INFO ]     output0 (node: output0) : f32 / [...] / [1,7,3..]
[Step 5/11] Resizing model to match image sizes and given batch
[ INFO ] Model batch size: 1
[ INFO ] Reshaping model: 'images': [1,3,640,640]
[ INFO ] Reshape model took 15.50 ms
[Step 6/11] Configuring input of the model
[ INFO ] Model inputs:
[ INFO ]     images (node: images) : u8 / [N,C,H,W] / [1,3,640,640]
[ INFO ] Model outputs:
[ INFO ]     output0 (node: output0) : f32 / [...] / [1,7,8400]
[Step 7/11] Loading the model to the device
[ INFO ] Compile model took 1574.51 ms
[Step 8/11] Querying optimal runtime parameters
[ INFO ] Model:
[ INFO ]   NETWORK_NAME: torch_jit
[ INFO ]   OPTIMAL_NUMBER_OF_INFER_REQUESTS: 16
[ INFO ]   NUM_STREAMS: 16
[ INFO ]   AFFINITY: Affinity.CORE
[ INFO ]   INFERENCE_NUM_THREADS: 16
[ INFO ]   PERF_COUNT: False
[ INFO ]   INFERENCE_PRECISION_HINT: <Type: 'float32'>
[ INFO ]   PERFORMANCE_HINT: PerformanceMode.THROUGHPUT
[ INFO ]   PERFORMANCE_HINT_NUM_REQUESTS: 0
[Step 9/11] Creating infer requests and preparing input tensors
[ WARNING ] No input files were given for input 'images'!. This input will be filled with random values!
[ INFO ] Fill input 'images' with random values
[Step 10/11] Measuring performance (Start inference asynchronously, 16 inference requests, limits: 15000 ms duration)
[ INFO ] Benchmarking in inference only mode (inputs filling are not included in measurement loop).
[ INFO ] First inference took 52.41 ms
[Step 11/11] Dumping statistics report
[ INFO ] Count:            2272 iterations
[ INFO ] Duration:         15104.41 ms
[ INFO ] Latency:
[ INFO ]    Median:        51.05 ms
[ INFO ]    Average:       105.75 ms
[ INFO ]    Min:           49.34 ms
[ INFO ]    Max:           727.25 ms
[ INFO ] Throughput:   150.42 FPS
```

**Comment by wangleis** [comment]
@dhaval-zala-aivid Looks like the result of benchmark app is good. Could you refer https://github.com/openvinotoolkit/openvino/blob/master/tools/benchmark_tool/openvino/tools/benchmark/main.py to refine your app or share your whole app?

**Comment by dhaval-zala-aivid** [comment]
@wangleis 
I have uploaded the app with the model [here](https://drive.google.com/drive/folders/1ta32_564Buz9sKnRv59QJblp6Ivaf7Bb?usp=sharing)
you can run `DvlEYE.py` that will run `yolov8AividEYE.py` to run the yolov8n model.

**Comment by wangleis** [comment]
@dhaval-zala-aivid Looks like the app is frame based loop which do inference one by one. 

Please refer async_inference_full_mode() in https://github.com/openvinotoolkit/openvino/blob/master/tools/benchmark_tool/openvino/tools/benchmark/benchmark.py to use infer_queue and data_queue for concurrency inference.

**Comment by divdaisymuffin** [comment]
@wangleis we have tried https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/115-async-api this as well but still on xeon cpu usage remains 50% only.
Please check if we are missing something in this, please correct, and help us to utilize proper CPU on xeon

**Comment by wangleis** [comment]
@divdaisymuffin  When hyper threading is enabled on two sockets Xeon platform, OpenVINO only use half processors for inference on throughput mode by default. 

If you want to use all processors in your case, please type below comment
`compiled_model = ie.compile_model(model=model, device_name="CPU", config={"CPU":"INFERENCE_NUM_THREADS 16"})`

**Comment by hbalasu1** [comment]
Hi @divdaisymuffin 

Could you let me know if your issue has been resolved?

**Comment by divdaisymuffin** [comment]
@hbalasu1 No its not, still we are not able to utilize full resources, and we cant deploy more instances.

**Comment by akashAD98** [comment]
https://github.com/openvinotoolkit/openvino_notebooks/issues/916
i want to add this issu,because  I'm also not able to utilize CPU

**Comment by wangleis** [comment]
@divdaisymuffin https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/115-async-api using 50% CPU on two sockets XEON platform is expected behavior by default. 

If you want to try 16 cores instead of 8 cores on your platform, please modify https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/115-async-api as below

```
# compile the model for the CPU (you can choose manually CPU, GPU, MYRIAD etc.)
# or let the engine choose the best available device (AUTO)
compiled_model = ie.compile_model(model=model, device_name="CPU", config={"INFERENCE_NUM_THREADS":"16"})
```

**Comment by dhaval-zala-aivid** [comment]
@wangleis we have tried this, you can see my script [here](https://drive.google.com/drive/folders/1ta32_564Buz9sKnRv59QJblp6Ivaf7Bb) but still its not able to utilize CPU.

**Comment by yury-gorbachev** [comment]
Judging by the code you are also benchmarking reading from file, image preprocessing and post-processing.
So, strictly speaking there is more than inference that is happening there, with reading being somewhat unpredictable from benchmarking standpoint.

**Comment by dhaval-zala-aivid** [comment]
@yury-gorbachev @wangleis 

I have removed all the image processing and just running the model only on Xeon by given code. Still its not able to use full CPU
```
from openvino.runtime import Core, Model
import openvino.runtime as ov
import time
import numpy as np

core = Core()
model_path = "/home/clouduser/Dhavalsinh/asynctesting/yolov8neno_msi2/msi2.xml"
ov_model = core.read_model(model_path)
ov_model.reshape([1,3, 416, 416])
input_layer_ir = ov_model.input(0)
config = {"PERFORMANCE_HINT": "THROUGHPUT","INFERENCE_NUM_THREADS": "16"}
device = "CPU"
compiled_model = core.compile_model(ov_model,device, config)
infer_request = compiled_model.create_infer_request()
frame_count = 0
start_time = time.time()
while True:
    image = np.random.rand(1,3,416,416).astype('float32')
    infer_request.set_tensor(input_layer_ir, ov.Tensor(image))
    infer_request.start_async()
    infer_request.wait()
    result = infer_request.get_output_tensor(0).data
    # print(result.shape)
    frame_count +=1 
    try:
          
        print("FPS: ",(frame_count/(int(time.time())-int(start_time))))
    except:
        pass
```

CPU performance:


![Screenshot from 2023-03-21 11-32-38](https://user-images.githubusercontent.com/91198866/226529851-d236053e-2204-4e4a-8b13-2a8b6efa90c2.png)
![Screenshot from 2023-03-21 11-33-47](https://user-images.githubusercontent.com/91198866/226529866-d6aabd59-c4b4-4af8-9ed4-62dd4440bec8.png)

**Comment by wangleis** [comment]
@dhaval-zala-aivid Throughput performance hint expect user user run inference with multiple inputs in parallel. However there is only one input during inference in your sample code. Please refer to AsyncInferQueue part in https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/115-async-api to perform parallel inference with optimal number of inference requests.

Xeon 6336Y is 24 cores 48 threads CPU. However your system show 2 sockets, 16 cores 16 threads CPU. May I know if you are running in VM or Docker? If yes, could you share what's the real CPU processors on bare metal machine? 

If you are running docker on 8 cores 16 threads CPU (hyperthreading enabled) but docker reports 16 cores 16 threads CPU, OpenVINO runtime will schedule inference for 16 cores 16 threads CPU and will not get full CPU utilization.

**Comment by dhaval-zala-aivid** [comment]
@wangleis 

I have tried with AsyncInferQueue by given code and played with different values of `QUEUE = 4` but I didn't find any improvement. I run [115-async-api](https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/115-async-api) but this also not utilising the CPU more than 50%.


```

from openvino.runtime import Core, Model
import openvino.runtime as ov
import time
import numpy as np
from openvino.runtime import Core, AsyncInferQueue

core = Core()
model_path = "yolov8neno.xml"

ov_model = core.read_model(model_path)
ov_model.reshape([1,3, 640, 640])
input_layer_ir = ov_model.input(0)
config = {"PERFORMANCE_HINT": "THROUGHPUT","NUM_STREAMS":"16"}
compiled_model = core.compile_model(ov_model,device, config)

device = "CPU"
QUEUE = 4
frame_count = 0
start_time = time.time()
infer_queue = AsyncInferQueue(compiled_model, QUEUE)

while True:
    image = np.random.rand(QUEUE,3,640,640).astype('float32')
    for i, input_tensor in enumerate(image):
        input_tensor = np.expand_dims(input_tensor, axis=0)
        infer_queue.start_async({input_layer_ir.any_name: input_tensor}) 
    infer_queue.wait_all()

    frame_count +=1 
    try:
          
        print("FPS: ",QUEUE*(frame_count/(int(time.time())-int(start_time))))
    except:
        pass
```

I'm running python script on VM given by one of cloud provider. They said the  Hyperthreading is enabled and because the servers are in shared infra, Bare metal machine details are confidential.

**Comment by wangleis** [comment]
@dhaval-zala-aivid Could you share more detail about "I run [115-async-api](https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/115-async-api) but this also not utilising the CPU more than 50%."?

There are 3 different CPU utilization situation about "not utilising the CPU more than 50%."

1. Utilization of half processors are more than 90%. Avg is not more than 50%. 
2. Utilization of only one processor are more than 90%. Avg is not more than 50%. 
3. Utilization of all processor are not more than 50%

There are three tests in  [115-async-api](https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/115-async-api).

- Test performance in Sync Mode
- Test the perfromance in Async Mode
- Test the perfromance with AsyncInferQueue

May I know what's the CPU utilization situation on each test? 

If all tests meet situation 3, it is VM environment issue, e.g., I/O bottleneck or memory bottleneck.

**Comment by ekurniaw** [comment]
@dhaval-zala-aivid, Xeon might be overkill for nano models. Could you try to reshape the model to process 100 and 1000 images (sample code below)? Also, do not enumerate the input for one image only. Try it using 100 and 1000 images. Then check the latency. The 100 images should be within 1 second as the benchmark_app reporting 150 FPS.

ov_model.reshape([100,3, 416, 416])
ov_model.reshape([1000,3, 416, 416])

**Comment by wangleis** [comment]
@akashAD98 May I know if there is any update?

**Comment by akashAD98** [comment]
@wangleis i tried asynchronusqueue & still im not able to achive the performace, im hardly getting 5,6 fps for yolov7x_int.xml model . i tried this repo 
https://github.com/OpenVINO-dev-contest/YOLOv7_OpenVINO_cpp-python

**Comment by akashAD98** [comment]
below is my script,here im loading the object detection model yolov7xint format model  & using this detector & passing it for tracking
```
from openvino.runtime import Core
from typing import List, Tuple, Dict
from utils.general import scale_coords, non_max_suppression
from openvino.runtime import Model
import numpy as np
import torch
from PIL import Image
from utils.datasets import letterbox
from pathlib import Path

core = Core()
# read converted model
#pat of object detector
model = core.read_model("object_detector/v7x_int8new.xml")


# load model on CPU device

model = core.compile_model(model, 'CPU')
# tried this also
setting = {"PERFORMANCE_HINT":"THROUGHPUT", "AFFINITY":"NUMA"}
model = core.compile_model(model=model, device_name="GPU/CPU", config=setting)

# this also
model = core.read_model("object_detector/yolov7_int8new.xml")
model = core.compile_model(model,config={"CPU":"INFERENCE_NUM_THREADS 32"})

def preprocess_image(img0: np.ndarray):
    """
    Preprocess image according to YOLOv7 input requirements. 
    Takes image in np.array format, resizes it to specific size using letterbox resize, converts color space from BGR (default in OpenCV) to RGB and changes data layout from HWC to CHW.
    
    Parameters:
      img0 (np.ndarray): image for preprocessing
    Returns:
      img (np.ndarray): image after preprocessing
      img0 (np.ndarray): original image
    """
    # resize
    img = letterbox(img0, auto=False)[0]
    
    # Convert
    img = img.transpose(2, 0, 1)
    img = np.ascontiguousarray(img)
    return img, img0


def prepare_input_tensor(image: np.ndarray):
    """
    Converts preprocessed image to tensor format according to YOLOv7 input requirements. 
    Takes image in np.array format with unit8 data in [0, 255] range and converts it to torch.Tensor object with float data in [0, 1] range
    
    Parameters:
      image (np.ndarray): image for conversion to tensor
    Returns:
      input_tensor (torch.Tensor): float tensor ready to use for YOLOv7 inference
    """
    input_tensor = image.astype(np.float32)  # uint8 to fp16/32
    input_tensor /= 255.0  # 0 - 255 to 0.0 - 1.0
    
    if input_tensor.ndim == 3:
        input_tensor = np.expand_dims(input_tensor, 0)
    return input_tensor
NAMES=[COCO80]

# colors for visualization
COLORS = {name: [np.random.randint(0, 255) for _ in range(3)]
          for i, name in enumerate(NAMES)}


def detect( image_path: Path, conf_thres: float = 0.50, iou_thres: float = 0.50, classes: List[int] = None, agnostic_nms: bool = False):
    """
    OpenVINO YOLOv7 model inference function. Reads image, preprocess it, runs model inference and postprocess results using NMS.
    Parameters:
        model (Model): OpenVINO compiled model.
        image_path (Path): input image path.
        conf_thres (float, *optional*, 0.25): minimal accpeted confidence for object filtering
        iou_thres (float, *optional*, 0.45): minimal overlap score for remloving objects duplicates in NMS
        classes (List[int], *optional*, None): labels for prediction filtering, if not provided all predicted labels will be used
        agnostic_nms (bool, *optiona*, False): apply class agnostinc NMS approach or not
    Returns:
       pred (List): list of detections with (n,6) shape, where n - number of detected boxes in format [x1, y1, x2, y2, score, label] 
       orig_img (np.ndarray): image before preprocessing, can be used for results visualization
       inpjut_shape (Tuple[int]): shape of model input tensor, can be used for output rescaling
    """
    output_blob = model.output(0)
    #img = np.array(Image.open(image_path))
    preprocessed_img, orig_img = preprocess_image(image_path)
    input_tensor = prepare_input_tensor(preprocessed_img)
    predictions = torch.from_numpy(model(input_tensor)[output_blob])
    pred = non_max_suppression(predictions, conf_thres, iou_thres, classes=classes, agnostic=agnostic_nms)
    boxes = []
    labels = []
    scores = []
    pred=pred[0]
    if len(pred):
        # Rescale boxes from input size to original image size
        pred[:, :4] = scale_coords(input_tensor.shape[2:], pred[:, :4], orig_img.shape).round()
        for *xyxy, conf, cls in pred:
            boxes.append(xyxy)
            labels.append(cls)
            scores.append(conf)
            print('class-OBJECTdetection',cls)
            #print('score_detection',conf)
            print('bbbox',xyxy)
    if len(boxes) == 0:
        boxes = np.array([]).reshape(0, 4)
        scores = np.array([])
        labels = np.array([])

    return np.array(boxes), np.array(scores), np.array(labels)

```

**Comment by wangleis** [comment]
@akashAD98 please try below configuration mentioned in https://github.com/openvinotoolkit/openvino_notebooks/issues/916 
```
setting = {"INFERENCE_NUM_THREADS":"32", "AFFINITY":"NUMA"}
compiled_model = core.compile_model(model=model, device_name="CPU", config=setting)
```

@dhaval-zala-aivid May I know if there is any update?

**Comment by wangleis** [comment]
Close this issue. Please feel free to reopen or create new issue if additional assistance is needed.

**Comment by dhaval-zala-aivid** [comment]
@wangleis As i have tried all approaches given in [115-async-api](https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/115-async-api)  the FPS are almost same in all the approach, its around 62 FPS. CPU utilisation is also same for all which is 50%. all threads of CPU are getting utilised but not a single thread is utilised  more than 50%.

**Comment by wangleis** [comment]
@dhaval-zala-aivid could you please add below commands at the end of [115-async-api](https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/115-async-api) and share the whole log of benchmark_app?

```
!benchmmark_app -m $model_path -d CPU -hint none -pin NUMA -nthreads 16
!benchmmark_app -m $model_path -d CPU -hint none -pin NUMA -nthreads 8
```

**Comment by wangleis** [comment]
@dhaval-zala-aivid May I know if there is any update?

**Comment by wangleis** [comment]
Close this issue. Please feel free to reopen or create new issue if additional assistance is needed.

**Comment by snehashis1997** [comment]
> hi,
> 
> I am trying to otimize CPU on my yolov6 nano model on openvino, I have found out couple of interesting things
> 
> When I am using i3 and i7, I am able to run 5 instances of same application on i3, 8 on i7, with 95% CPU usage, while on Xeon silver (6 instances) and Xeon gold only 4, and CPu goes upto 50% max on both Xeon, which I think they have more capacity but somehow they are not taking CPU.
> 
> Can you help me with this, because as per Openvino benchmarking Xeon should work better, but in my observation its opposite,
> 
> Thanks

Hi @dhaval-zala-aivid , did you benchmark the yolov6 int8 openvino model in intel CPU or is it fp32 IR model?
